File: atividade-8/atividade.py
class AFN:
    def __init__(self):
        self.states = {'q0', 'q1', 'q2'}
        self.start_state = 'q0'
        self.accept_states = {'q0'}

    def is_accepted(self, string):
        current_states = {self.start_state}  

        for char in string:
            next_states = set()
            for state in current_states:
                if state == 'q0':
                    if char == '0':
                        next_states.add('q1')  
                    else:
                        next_states.add('q0')  
                elif state == 'q1':
                    if char == '0':
                        next_states.add('q2')  
                    else:
                        next_states.add('q1')  
                elif state == 'q2':
                    if char == '0':
                        next_states.add('q0')  
                    else:
                        next_states.add('q2')  

            current_states = next_states

        return any(state in self.accept_states for state in current_states)

if __name__ == '__main__':
    afn = AFN()
    print(afn.is_accepted("000"))    
    print(afn.is_accepted("001"))    
    print(afn.is_accepted("010"))    
    print(afn.is_accepted("111"))    
    print(afn.is_accepted("0001"))   
    print(afn.is_accepted("0000"))   

File: atividade-8/atividade.go
package main

import (
	"fmt"
)

type AFN struct {
	states       map[string]bool
	startState   string
	acceptStates map[string]bool
}

func NewAFN() *AFN {
	return &AFN{
		states:       map[string]bool{"q0": true, "q1": true, "q2": true},
		startState:   "q0",
		acceptStates: map[string]bool{"q0": true},
	}
}

func (afd *AFN) IsAccepted(input string) bool {
	currentStates := map[string]bool{afd.startState: true}

	for _, char := range input {
		nextStates := make(map[string]bool)
		for state := range currentStates {
			if state == "q0" {
				if char == '0' {
					nextStates["q1"] = true
				} else {
					nextStates["q0"] = true
				}
			} else if state == "q1" {
				if char == '0' {
					nextStates["q2"] = true
				} else {
					nextStates["q1"] = true
				}
			} else if state == "q2" {
				if char == '0' {
					nextStates["q0"] = true
				} else {
					nextStates["q2"] = true
				}
			}
		}
		currentStates = nextStates
	}

	for state := range currentStates {
		if afd.acceptStates[state] {
			return true
		}
	}
	return false
}

func main() {
	afd := NewAFN()
	fmt.Println(afd.IsAccepted("000")) 
	fmt.Println(afd.IsAccepted("001"))  
	fmt.Println(afd.IsAccepted("010"))  
	fmt.Println(afd.IsAccepted("111"))  
	fmt.Println(afd.IsAccepted("0001")) 
	fmt.Println(afd.IsAccepted("0000")) 
}
